ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -389,38 +221,215 @@ nullNetworkServerTracers = NetworkServerTracers {
+  -> IO t
+withConnections mk = Connection.concurrent (connection mk)
+
+-- | Handle any connection (remotely- or locally-initiated).
+-- After filling in the first 7 parameters, you get a function that can be
+-- used to create a `Connections.Concurrent.concurrent` `Connections` term.
+--
+-- From that, you can derive a socket accept loop by using `acceptLoopOn`.
+-- Outgoing connections can be made using `runClientWith` on a pair of
+-- socket addresses, and referencing that `Connections` term.
+connection
+  :: forall ptcl vNumber provenance reject request.
+     ( ProtocolEnum ptcl
+     , MiniProtocolLimits ptcl
+     , Ord ptcl
+     , Enum ptcl
+     , Bounded ptcl
+     , Ord vNumber
+     , Enum vNumber
+     , Serialise vNumber
+     , Typeable vNumber
+     , Show vNumber
+     , Show ptcl
+     )
+  => (forall provenance'. request provenance' -> ConnectionData ptcl vNumber provenance')
+  -> Initiated provenance
+  -> Connections.ConnectionId
+  -> Socket.Socket
+  -> request provenance
+  -> IO (Connection.Decision IO provenance reject (ConnectionHandle IO))
+connection mk _ connid socket request = case mk request of
+
+    -- TODO should take an error policy. We'll use it for exception handling,
+    -- to figure out when to blow everything up.
+    ConnectionDataLocal tracers vCodec versions ->
+        outgoingConnection vCodec tracers versions connid socket
+
+    ConnectionDataRemote tracers errPolicies vCodec accept versions ->
+        incomingConnection tracers vCodec accept versions errPolicies connid socket
+
+-- | What to do for outgoing (locally-initiated) connections.
+outgoingConnection
+  :: forall ptcl vNumber vDataT appType reject a b.
+     ( Mx.HasInitiator appType ~ True
+     , ProtocolEnum ptcl
+     , Ord ptcl
+     , Enum ptcl
+     , Bounded ptcl
+     , Ord vNumber
+     , Enum vNumber
+     , Serialise vNumber
+     , Typeable vNumber
+     , Show vNumber
+     , Show ptcl
+     , MiniProtocolLimits ptcl
+     )
+  => VersionDataCodec vDataT CBOR.Term
+  -> NetworkConnectTracers ptcl vNumber
+  -> Versions vNumber vDataT (OuroborosApplication appType ConnectionId ptcl IO BL.ByteString a b)
+  -- ^ application to run over the connection
+  -> Connections.ConnectionId
+  -> Socket.Socket       -- ^ Socket to peer; could have been established by us or them.
+  -> IO (Connection.Decision IO Local reject (ConnectionHandle IO))
+outgoingConnection versionDataCodec tracers versions connId sd =
+    -- Always accept and run initiator mode mux on the socket.
+    pure $ Connection.Accept $ \_connThread -> do
+        statusVar <- atomically (newTVar Running)
+        let connectionHandle = ConnectionHandle
+              { status = readTVar statusVar }
+            action = mask $ \restore -> do
+              restore (runInitiator versionDataCodec tracers versions connId sd) `catch`
+                (\e -> atomically (writeTVar statusVar (Finished (Just e))))
```

Comment:
`e` might be an un-evaluated thunk, using `Control.Monad.Class.MonadSTM.Strict` (and the suggested `ConnectionStatus`) would fix that.

Reply:
Is this a big deal? It's an exception, so almost surely not a very big thunk, right? And for how long does it live? Not very long I think: we finish that transaction to write `Finished` and then throw it.

If the caller leaks handles to these connections, that's a whole other story. Forcing this exception wouldn't make a difference.